### PR TITLE
bpo-35975: Support parsing earlier minor versions of Python 3

### PR DESCRIPTION
**Pull request by gvanrossum**
This adds a `feature_version` flag to `ast.parse()` (documented) and `compile()` (hidden) that allow tweaking the parser to support older versions of the grammar. In particular if `feature_version` is 5 or 6, the hacks for the `async` and `await` keyword from PEP 492 are reinstated. (For 7 or higher, these are unconditionally treated as keywords, but they are still special tokens rather than `NAME` tokens that the parser driver recognizes.)

https://bugs.python.org/issue35975

### CODE REVIEW

**Comment by gvanrossum** [comment]
I could use some help finding the cause of the one remaining test failure -- test_parser.py computes the size of tree nodes differently somehow.

**Comment by gvanrossum** [comment]
The main thing left to do is adding tests. There are some tests in typed_ast that I can migrate over. But first I think we should complete the debate over on [bpo](https://bugs.python.org/issue35975) about whether this is desirable at all.

File: Lib/ast.py [hunk 1]
```diff
@@ -36,7 +37,11 @@ def parse(source, filename='<unknown>', mode='exec', *, type_comments=False):
     flags = PyCF_ONLY_AST
     if type_comments:
         flags |= PyCF_TYPE_COMMENTS
-    return compile(source, filename, mode, flags)
+    return compile(source, filename, mode, flags,
+                   dont_inherit=False,
+                   optimize=-1,
```
Comment (ilevkivskyi) [inline Lib/ast.py]: Why we now need to pass these two extra arguments? Maybe add a comment?

File: Python/ast.c [hunk 2]
```diff
@@ -2053,15 +2068,21 @@ ast_for_comprehension(struct compiling *c, const node *n)
...
         }
         else {
             sync_n = CHILD(n, 0);
         }
         REQ(sync_n, sync_comp_for);
 
+        /* Async comprehensions only allowed in Python 3.6 and greater */
+        if (is_async && c->c_feature_version < 6) {
+            ast_error(c, n,
+                    "Async comprehensions are only supported in Python 3.6 and greater");
```
Comment (ilevkivskyi) [inline Python/ast.c]: Indentation here and in several other places is a bit unusual, I would rather justify it after `(`.

File: Python/ast.c [hunk 3]
```diff
@@ -1746,6 +1758,12 @@ ast_for_funcdef_impl(struct compiling *c, const node *n0,
     node *tc;
     string type_comment = NULL;
 
+    if (is_async && c->c_feature_version < 5) {
+        ast_error(c, n,
+                  "Async functions are only supported in Python 3.5 and greater");
```
Comment (ilevkivskyi) [inline Python/ast.c:1736]: Here indentation looks good.

**Review by ilevkivskyi — APPROVED** [review]
Thanks! This looks very clean, I just have few minor suggestions.

(I also there is a merge conflict now.)

File: Lib/ast.py (hunk 1 above)
Comment (gvanrossum) [reply Lib/ast.py]: Hm, those are the default values that I somehow copied when cleaning this up. I'll remove them -- we just need `feature_version`.

File: Python/ast.c (hunk 2 above)
Comment (gvanrossum) [reply Python/ast.c]: OK, fixed. My guess is that at some point in the past that function had had a name that was 2 characters shorter. :-)

**Comment by gvanrossum** [comment]
OK, this is ready for final review and merge. (Sorry that the rebase lost some of the review history, I'm used to different tooling.)

**Comment by gvanrossum** [comment]
(Well, I promised tests. Upcoming. Docs are already done.)

File: Doc/library/ast.rst [hunk 4]
```diff
@@ -145,13 +145,18 @@ and classes for traversing abstract syntax trees:
    modified to correspond to :pep:`484` "signature type comments",
    e.g. ``(str, int) -> List[str]``.
 
+   Also, setting ``feature_version`` to the minor version of an
+   earlier Python 3 version will attempt to parse using that version's
+   grammar.  For example, setting ``feature_version=4`` will allow
```
Comment (ilevkivskyi) [inline Doc/library/ast.rst:150]: Maybe add that `4` is the lowest supported value?

File: Lib/test/test_type_comments.py [hunk 5]
```diff
@@ -155,80 +175,113 @@ def favk(
...
-    def parse(self, source):
-        return ast.parse(source, type_comments=True)
+    lowest = 4  # Lowest minor version supported
+    highest = sys.version_info[1]  # Highest minor version
+
+    def parse(self, source, feature_version=highest):
+        return ast.parse(source, type_comments=True,
+                         feature_version=feature_version)
+
+    def parses(self, source, minver=lowest, maxver=highest, expected_regex=""):
```
Comment (ilevkivskyi) [inline Lib/test/test_type_comments.py]: `parses` is not very descriptive and is easy to confuse with `parse`, maybe `parse_all`, or `parse_all_versions`?

File: Lib/test/test_type_comments.py [hunk 6]
```diff
@@ -20,6 +21,25 @@ def bar():  # type: () -> None
...
+"""
+
+matmul = """\
+a = b @ c
+"""
+
+fstring = """\
+a = 42
+f"{a}"
+"""
```
Comment (ilevkivskyi) [inline Lib/test/test_type_comments.py:41]: Maybe add underscores in numeric literals for completeness?

**Review by ilevkivskyi — APPROVED** [review]
Thanks for the updates! All looks good, I have few optional suggestions.

**Pull request merged**